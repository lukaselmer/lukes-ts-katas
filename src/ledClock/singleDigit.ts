import { truthy } from '../utils/array'

export function singleDigit(num: number) {
  if (num < 0 || num > 9) throw new Error(`Invalid digit: ${num}`)
  return calculateSplitDigits()[num]
}

function calculateSplitDigits() {
  const lines = rawDigits.split('\n').filter(truthy)
  const height = lines.length / 10
  const singleDigits: string[][] = []
  lines.forEach((line, index) => {
    const digitIndex = Math.floor(index / height)
    singleDigits[digitIndex] = singleDigits[digitIndex] || []
    singleDigits[digitIndex].push(line)
  })
  return singleDigits
}

const rawDigits = `
╔══╗
║  ║
╚══╝
╗
║
╝
╔══╗
╔══╝
╚══╝
══╗
 ═╣
══╝
╔  ╗
╚══╣
   ╝
╔══╗
╚══╗
╚══╝
╔══╗
╠══╗
╚══╝
╔══╗
   ║
   ╝
╔══╗
╠══╣
╚══╝
╔══╗
╚══╣
╚══╝
`
