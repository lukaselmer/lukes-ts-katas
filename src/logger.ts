import { assertNever } from 'assert-never'

let logger: undefined | 'real' | 'fake'
const fakeLogEntries: string[][] = []

export function initLogger(type: 'real' | 'fake') {
  if (logger) throw new Error('logger already set')
  logger = type
}

export function discardLogs() {
  fakeLogEntries.splice(0, fakeLogEntries.length)
}

export function fakeLogs() {
  if (logger !== 'fake') throw new Error('logger not set to fake')

  return fakeLogEntries.map((entry) => entry.join(', ')).join('\n')
}

export function consoleLog(...args: string[]) {
  if (!logger) throw new Error('logger not set')
  else if (logger === 'real') console.log(...args)
  else if (logger === 'fake') fakeLogEntries.push(args)
  else assertNever(logger)
}
