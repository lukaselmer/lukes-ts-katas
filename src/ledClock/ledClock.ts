import { consoleLog } from '../logger'

export function ledClock(value: number) {
  if (value === 1) {
    consoleLog('')
    consoleLog('   ╗')
    consoleLog('   ║')
    consoleLog('   ╝')
    consoleLog('')
  }
  if (value === 2) {
    consoleLog('')
    consoleLog('╔══╗')
    consoleLog('╔══╝')
    consoleLog('╚══╝')
    consoleLog('')
  }
  if (value === 3) {
    consoleLog('')
    consoleLog(' ══╗')
    consoleLog('  ═╣')
    consoleLog(' ══╝')
    consoleLog('')
  }
  if (value === 4) {
    consoleLog('')
    consoleLog('╔  ╗')
    consoleLog('╚══╣')
    consoleLog('   ╝')
    consoleLog('')
  }
  if (value === 5) {
    consoleLog('')
    consoleLog('╔══╗')
    consoleLog('╚══╗')
    consoleLog('╚══╝')
    consoleLog('')
  }
  if (value === 6) {
    consoleLog('')
    consoleLog('╔══╗')
    consoleLog('╠══╗')
    consoleLog('╚══╝')
    consoleLog('')
  }
  if (value === 7) {
    consoleLog('')
    consoleLog('╔══╗')
    consoleLog('   ║')
    consoleLog('   ╝')
    consoleLog('')
  }
  if (value === 8) {
    consoleLog('')
    consoleLog('╔══╗')
    consoleLog('╠══╣')
    consoleLog('╚══╝')
    consoleLog('')
  }
  if (value === 9) {
    consoleLog('')
    consoleLog('╔══╗')
    consoleLog('╚══╣')
    consoleLog('╚══╝')
    consoleLog('')
  }
}

function splitDigits() {
  console.log(rawDigits.split('\n'))
}

const rawDigits = `
  ╗
  ║
  ╝
╔══╗
╔══╝
╚══╝
══╗
 ═╣
══╝
╔  ╗
╚══╣
   ╝
╔══╗
╚══╗
╚══╝
╔══╗
╠══╗
╚══╝
╔══╗
   ║
   ╝
╔══╗
╠══╣
╚══╝
╔══╗
╚══╣
╚══╝
`

splitDigits()
