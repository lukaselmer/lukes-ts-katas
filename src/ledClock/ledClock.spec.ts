import { beforeEach, beforeAll, describe, expect, it } from 'vitest'
import { discardLogs, fakeLogs, initLogger } from '../logger'
import { ledClock } from './ledClock'

beforeAll(() => initLogger('fake'))
beforeEach(discardLogs)

describe('ledClock', () => {
  describe('single digits', () => {
    it.each([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])('generates the correct output for digit %s', (num) => {
      ledClock(num)
      expect(fakeLogs()).toMatchSnapshot()
    })

    it('generates the correct output for digit 3', () => {
      expect(fakeLogs()).toMatchInlineSnapshot('""')
      ledClock(3)
      expect(fakeLogs()).toMatchInlineSnapshot(`
      "
       ══╗
        ═╣
       ══╝
      "
    `)
    })

    it('generates the correct output for digit 8', () => {
      expect(fakeLogs()).toMatchInlineSnapshot('""')
      ledClock(8)
      expect(fakeLogs()).toMatchInlineSnapshot(`
      "
      ╔══╗
      ╠══╣
      ╚══╝
      "
    `)
    })
  })

  it('generates the correct output for digit 351', () => {
    expect(fakeLogs()).toMatchInlineSnapshot('""')
    ledClock(351)
    expect(fakeLogs()).toMatchInlineSnapshot(`
      "
       ══╗ ╔══╗    ╗
        ═╣ ╚══╗    ║
       ══╝ ╚══╝    ╝
      "
    `)
  })
})
