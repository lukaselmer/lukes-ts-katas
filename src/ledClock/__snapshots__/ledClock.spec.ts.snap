// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ledClock > single digits > generates the correct output for digit 0 1`] = `""`;

exports[`ledClock > single digits > generates the correct output for digit 1 1`] = `""`;

exports[`ledClock > single digits > generates the correct output for digit 2 1`] = `""`;

exports[`ledClock > single digits > generates the correct output for digit 3 1`] = `
"
 ══╗
  ═╣
 ══╝
"
`;

exports[`ledClock > single digits > generates the correct output for digit 4 1`] = `""`;

exports[`ledClock > single digits > generates the correct output for digit 5 1`] = `""`;

exports[`ledClock > single digits > generates the correct output for digit 6 1`] = `""`;

exports[`ledClock > single digits > generates the correct output for digit 7 1`] = `""`;

exports[`ledClock > single digits > generates the correct output for digit 8 1`] = `
"
╔══╗
╠══╣
╚══╝
"
`;

exports[`ledClock > single digits > generates the correct output for digit 9 1`] = `""`;
